File: python/nmti/root_finding/newton.py


def newton(f, df, x0, tol=1e-10, max_iter=100):
    """
    Newton's method for root finding.

    Parameters
    ----------
    f : callable
        Function whose root is sought.
    df : callable
        Derivative of f.
    x0 : float
        Initial guess.
    tol : float, optional
        Convergence tolerance.
    max_iter : int, optional
        Maximum number of iterations.

    Returns
    -------
    x : float
        Estimated root.
    n_iter : int
        Number of iterations performed.
    """
    x = x0
    for i in range(max_iter):
        fx = f(x)
        dfx = df(x)
        if dfx == 0:
            raise ZeroDivisionError("Derivative is zero, Newton method will fail.")
        x_new = x - fx / dfx
        if abs(x_new - x) < tol:
            return x_new, i + 1
        x = x_new

    raise RuntimeError("Newton method did not converge with max_iter.")

File: fortran/src/root_finding/newton.f90
module newton_mod
  implicit none
contains

  function newton(f, df, x0, tol, max_iter, iterations) result(root)
    !
    ! Newton's method for root finding.
    !
    interface
       real(kind=8) function f(x)
         real(kind=8), intent(in) :: x
       end function f

       real(kind=8) function df(x)
         real(kind=8), intent(in) :: x
       end function df
    end interface

    real(kind=8), intent(in) :: x0, tol
    integer, intent(in) :: max_iter
    integer, intent(out) :: iterations
    real(kind=8) :: root, x, x_new, fx, dfx

    x = x0

    do iterations = 1, max_iter
       fx = f(x)
       dfx = df(x)

       if (dfx == 0.0d0) then
          print *, "Derivative is zero, Newton method will fail."
          stop
       end if

       x_new = x - fx / dfx

       if (abs(x_new - x) < tol) then
          root = x_new
          return
       end if

       x = x_new
    end do

    print *, "Newton did not converge with max_iter."
    stop

  end function newton


end module newton_mod
